flang/runtime/execute.cpp

```diff
@@ -64,22 +64,40 @@ void CheckAndStoreIntToDescriptor(
 // the CMDSTAT variable is not present, error termination is initiated.
 int TerminationCheck(int status, const Descriptor *cmdstat,
     const Descriptor *cmdmsg, Terminator &terminator) {
-  if (status == -1) {
-    if (!cmdstat) {
-      terminator.Crash("Execution error with system status code: %d", status);
-    } else {
-      StoreIntToDescriptor(cmdstat, EXECL_ERR, terminator);
-      CheckAndCopyCharsToDescriptor(cmdmsg, "Execution error");
-    }
-  }
 #ifdef _WIN32
   // On WIN32 API std::system returns exit status directly
   int exitStatusVal{status};
-  if (exitStatusVal == 1) {
 #else
   int exitStatusVal{WEXITSTATUS(status)};
-  if (exitStatusVal == 127 || exitStatusVal == 126) {
+  if (exitStatusVal == 1) {
+    if (!cmdstat) {
+      terminator.Crash("General Error with exit status code: 1");
+    } else {
+      StoreIntToDescriptor(cmdstat, INVALID_CL_ERR, terminator);
+      CheckAndCopyCharsToDescriptor(cmdmsg,
+          "General Error: a catch-all exit code for a variety of general "
+          "errors.");
+    }
+  } else if (exitStatusVal == 126) {
+    if (!cmdstat) {
+      terminator.Crash("Command cannot execute with exit status code: 126");
+    } else {
+      StoreIntToDescriptor(cmdstat, INVALID_CL_ERR, terminator);
+      CheckAndCopyCharsToDescriptor(cmdmsg,
+          "Command cannot execute: command was found, but it could not be "
+          "executed.");
+    }
+  } else if (exitStatusVal == 127) {
+    if (!cmdstat) {
+      terminator.Crash("Command not found with exit status code: 127");
+    } else {
+      StoreIntToDescriptor(cmdstat, INVALID_CL_ERR, terminator);
+      CheckAndCopyCharsToDescriptor(cmdmsg,
+          "Command not found: command was not found in the system's PATH");
+    }
+  } else
 #endif
+  if (exitStatusVal != 0) {
```

Comment:
I don't know if there is a prettier way to write this logic:
On Linux (remember the !=0 covers all other error exit code)
```
if exit ==1
else if exit == 126
else if exit == 127
else if exit != 0
```
On Windows
```
if exit != 0
```

So in my code it becomes:
```
#ifdef Linux
  if exit ==1
  else if exit == 126
  else if exit == 127
  else (connect to the following if to beocome a else-if)
#endif 
if exit != 0

```

Reply:
Thanks for the update!
It seems to me the "old" logic is

```
if (status == -1) { //  Case 1: common for both Linux and Windows
...
}
#ifdef windows
  if (exitStatusVal == 1) { // Case 2 for all errors on windows.
#else
  if (exitStatusVal == 126 or 127 { // Case 3 (126) and Case 4 (127) for Linux.
#end if
...
}
```

I think what is missing is there is no checking for `exitStatusVal == 1` to catch the case in my issue on Linux. 
I am not sure if we need `exitStatusVal != 0` case though. 
Based on your commits, may I suggest something like:
```
enum CMD_STAT {
  ASYNC_NO_SUPPORT_ERR = -2,
  NO_SUPPORT_ERR = -1,
  CMD_EXECUTED = 0,
  FORK_ERR = 1,
  COMMAND_EXECUTION_ERR = 2, // exit code 1
  COMMAND_CANNOT_EXECUTE_ERR = 3, // Linux exit code 126
  COMMAND_NOT_FOUND_ERR = 4, // Linux exit code 127
  SIGNAL_ERR = 5
};
...

  if (status == -1) {
    if (!cmdstat) {
      terminator.Crash("Command line execution is not supported with system status code: %d.", status);
    } else {
      StoreIntToDescriptor(cmdstat, NO_SUPPORT_ERR, terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg, "Command line execution is not supported.");
    }
  }
#ifdef _WIN32
  // On WIN32 API std::system returns exit status directly
  int exitStatusVal{status};
  if (exitStatusVal == 1) {
    if (!cmdstat) {
      terminator.Crash(
          "Invalid command quit with exit status code: %d.", exitStatusVal);
    } else {
      StoreIntToDescriptor(cmdstat, INVALID_CL_ERR, terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg, "Invalid command line."); // Since Windows doesn't have different return code for different errors, use a generic error message.
    }
  }
#else
  int exitStatusVal{WEXITSTATUS(status)};
  if (exitStatusVal == 1) {
    if (!cmdstat) {
      terminator.Crash("Command line execution failed with exit code: 1.");
    } else {
      StoreIntToDescriptor(cmdstat, COMMAND_EXECUTION_ERR , terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg,
          "Command line execution failed with exit code: 1.");
    }
  } else if (exitStatusVal == 126) {
    if (!cmdstat) {
      terminator.Crash("Command cannot be executed with exit code: 126.");
    } else {
      StoreIntToDescriptor(cmdstat, COMMAND_CANNOT_EXECUTE_ERR, terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg,
          "Command cannot be executed with exit code: 126.");;
    }
  } else if (exitStatusVal == 127) {
    if (!cmdstat) {
      terminator.Crash("Command not found with exit code: 127.");
    } else {
      StoreIntToDescriptor(cmdstat, COMMAND_NOT_FOUND_ERR, terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg,
          "Command not found with exit code: 127.");
    }
  }
  ```
  
  Please note that I am not sure what case `status == -1` covers. I assumed it is for the case that a processor that doesn't support command line execution, so I changed the `CMD_STAT` of it from `EXECL_ERR` to `NO_SUPPORT_ERR`. Please correct me if I am wrong. BTW, 'NO_SUPPORT_ERR` seems is not used anywhere.

Reply:
Thanks a lot for your inputs, I agree that give specific `CMD_STAT` for common exit code is a good idea.

---

`exitStatusVal != 0` was meant to be a general error message, covers all other error code that wasn't specified by the if-else. Otherwise `CMD_MSG` and `CMD_STAT` would only be set if error code is 1, 126, 127.

---

>BTW, `NO_SUPPORT_ERR` seems is not used anywhere otherwise.  

On Windows
>https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/system-wsystem?view=msvc-170
>A return value of -1 indicates an error, and errno is set to one of the following values  

| Value | Description |
| --- | --- |
| E2BIG | The argument list (which is system-dependent) is too large. |
| ENOENT | The command interpreter can't be found. |
| ENOEXEC | The command-interpreter file can't be executed because the format isn't valid. |
| ENOMEM | Not enough memory is available to execute command; or available memory has been corrupted; or a non-valid block exists, which indicates that the calling process has been allocated incorrectly. |

On Linux: (`std::system` is achieved by `fork()`)
>https://man7.org/linux/man-pages/man3/system.3.html
>If a child process could not be created, or its status could not be retrieved, the return value is -1 and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to indicate the error.

In other words, a return code of -1 could mean `EXECL_ERR` **or** `NO_SUPPORT_ERR` based on the value of errno, thats why I used `EXECL_ERR`, for safety. On a second thought, I think it would be best to report the value of `errno` as well.

Reply:
Ok, I see.
1. For `status == -1` case, I am OK with specifying specific exit code with corresponding error messages depends on the platform.
2. For `exitStatusVal == 1/126/127` case, we have already covered them.
3. For `exitStatusVal != 0` on top of 1 and 2, I am OK with adding another `else` to cover all "other" cases for all platform.

Reply:
Thanks a lot! I have updated the code based on previous comments.
I have rebased on main, hopefully this will solve some unrelated test failure.... Waiting for CI result.